File: mergeTwoBinaryTrees.swift
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init(_ val: Int) {
 *         self.val = val
 *         self.left = nil
 *         self.right = nil
 *     }
 * }
 */
class Solution {
    func mergeTrees(_ t1: TreeNode?, _ t2: TreeNode?) -> TreeNode? {
        if t1 == nil && t2 == nil {
            return nil
        }
        let root = TreeNode(0)
        _mergeTrees(root, t1, t2)
        return root
    }
    func _mergeTrees(_ root: TreeNode, _ t1: TreeNode?, _ t2: TreeNode?) {
        if t1 != nil {
            root.val = t1!.val
        }
        if t2 != nil {
            root.val += t2!.val
        }
        if t1?.left != nil || t2?.left != nil {
            root.left = TreeNode(0)
            _mergeTrees(root.left!, t1?.left, t2?.left)
        }
        if t1?.right != nil || t2?.right != nil {
            root.right = TreeNode(0)
            _mergeTrees(root.right!, t1?.right, t2?.right)
        }
    }
}

File: LeeCode/mergeTwoBinaryTrees.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def mergeTreesImp(self, root: TreeNode, t1: TreeNode, t2: TreeNode):
        t1l = None
        t1r = None
        t2l = None
        t2r = None
        if t1:
            root.val = t1.val
            t1l = t1.left
            t1r = t1.right
        if t2:
            root.val += t2.val
            t2l = t2.left
            t2r = t2.right
        if t1l or t2l:
            lNode = TreeNode(0)
            root.left = lNode
            self.mergeTreesImp(lNode, t1l, t2l)
        if t1r or t2r:
            rNode = TreeNode(0)
            root.right = rNode
            self.mergeTreesImp(rNode, t1r, t2r)

    def mergeTrees(self, t1: TreeNode, t2: TreeNode) -> TreeNode:
        if t1 or t2:
            root = TreeNode(0)
            self.mergeTreesImp(root, t1, t2)
            return root
        else:
            return None
        
